src/installer/BeatPackageCompiler/AgentCustomAction.cs

```diff
@@ -0,0 +1,82 @@
+﻿using System;
+using System.Diagnostics;
+using System.IO;
+using Microsoft.Deployment.WindowsInstaller;
+
+namespace Elastic.PackageCompiler.Beats
+{
+    public class AgentCustomAction
+    {
+        [CustomAction]
+        public static ActionResult InstallAction(Session session)
+        {
+            try
+            {
+                // If there are no install args, we stop here
+                // // (the MSI just copied the files, there will be no agent-install)
+                if (string.IsNullOrEmpty(session["INSTALLARGS"]))
+                    return ActionResult.Success;
+
+                string install_args = session["INSTALLARGS"];
+                string install_folder = Path.Combine(session["INSTALLDIR"], session["exe_folder"]);
+                System.Diagnostics.Process process = new System.Diagnostics.Process();
+                process.StartInfo.WorkingDirectory = install_folder;
+                process.StartInfo.FileName = "elastic-agent.exe";
+                process.StartInfo.Arguments = "install -f " + install_args;
+                process.StartInfo.CreateNoWindow = true;
+                process.Start();
+                process.WaitForExit();
+
+                session.Log("Agent install return code:" + process.ExitCode);
+
+                if (process.ExitCode == 0)
+                {
+                    // If agent got installed properly, we can go ahead and remove all the files installed by the MSI (best effort)
+                    new DirectoryInfo(install_folder).Delete(true);
```

Comment:
This is marked as best effort but I think if it can't delete a file/folder the MSI, this will throw an exception which will fail the customaction with a generic 1603 error and a log that looks like this:
```
Agent install return code: 0
CustomAction InstallAction returned actual error code 1603...
```
and the MSI will report as having ended prematurely.

I am able to reproduce this locally by opening the beats folder with explorer during install.

Reply:
ack. but you do agree that this code should exist, right? i just need to wrap it in a dedicated exception handler..